compiler/rustc_macros/src/symbols.rs

```diff
@@ -151,51 +149,99 @@ fn symbols_with_errors(input: TokenStream) -> (TokenStream, Vec<syn::Error>) {
+
     // Generate the listed keywords.
+    let mut keyword_stream = quote! {};
     for keyword in input.keywords.iter() {
         let name = &keyword.name;
         let value = &keyword.value;
         let value_string = value.value();
-        check_dup(keyword.name.span(), &value_string, &mut errors);
-        prefill_stream.extend(quote! {
-            #value,
-        });
+        let symbol_index = if check_dup(keyword.name.span(), &value_string, &mut errors).is_ok() {
+            let symbol_index = symbol_strings.len() as u32;
+            symbol_strings.push(value_string);
+            symbol_index
+        } else {
+            // Compilation is going to fail, because of a duplicate symbol name.
+            // Add a symbol definition, even one with a bogus value, to minimize
+            // unrelated errors.
+            0
+        };
         keyword_stream.extend(quote! {
-            pub const #name: Symbol = Symbol::new(#counter);
+            pub const #name: Symbol = Symbol::new(#symbol_index);
         });
-        counter += 1;
     }
 
+    // Generate symbols for the strings "0", "1", ..., "9".
+    let digits_base = symbol_strings.len() as u32;
+    for n in 0..10 {
+        let n_string = n.to_string();
+        if check_dup(Span::call_site(), &n_string, &mut errors).is_ok() {
+            symbol_strings.push(n_string);
+        }
+    }
+
+    let is_digit_symbol = |s: &str| -> Option<u32> {
+        // Comparing len() to 1 is valid; all single-byte strings consist of a single
+        // ASCII character.
+        if s.len() == 1 {
+            let c = s.chars().next().unwrap();
+            if c >= '0' && c <= '9' {
+                return Some(c as u32 - '0' as u32 + digits_base);
+            }
+        }
+        None
+    };
+
     // Generate the listed symbols.
+    let mut symbols_stream = quote! {};
     for symbol in input.symbols.iter() {
         let name = &symbol.name;
+        let name_string = symbol.name.to_string();
+        check_order(symbol.name.span(), &name_string, &mut errors);
         let value = match &symbol.value {
             Some(value) => value.value(),
-            None => name.to_string(),
+            None => name_string,
+        };
+
+        let symbol_index = if let Some(i) = is_digit_symbol(&value) {
+            // We allow duplicates for digits, because we are not creating
+            // a new symbol index, just a new `sym::foo` name for an existing symbol.
+            i
+        } else {
+            if check_dup(symbol.name.span(), &value, &mut errors).is_ok() {
+                let symbol_index = symbol_strings.len() as u32;
+                symbol_strings.push(value);
+                symbol_index
+            } else {
+                // Compilation is going to fail, because of a duplicate symbol name.
+                // Add a symbol definition, even one with a bogus value, to minimize
+                // unrelated errors.
+                0
```

Comment:
If any value works here, then `symbol_index` would work too, so it can be moved out of the `check_dup` condition.

Reply:
Done.